### ISSUE TITLE
RequestCaptureStream does not write triggering record

### ISSUE THREAD

**Issue by gerrard00** [restify/node-restify#958]
I'm trying to use RequestCaptureStream with my restify app. I need to write to multiple streams and I would like the escalation feature of RequestCaptureStream to apply to all of those streams. Unfortunately, I haven't been able to wrap my head around the correct way to accomplish this.

Here's a simplified example of what I'm trying to do:

```js
'use strict';

let bunyan = require('bunyan');
let restify = require('restify');
let streams = [];
// the next two arrays are just for testing
let streamsWithCapture = [];
let streamsWithCaptureFirst = [];

streams.push({
  stream: process.stdout,
  level: 'info' 
});

streams.push({
  level: 'info',
  stream: require('fs').createWriteStream('/var/tmp/debug.log')
});

let requestCaptureStream = {
  level: bunyan.DEBUG,
  type: 'raw',
  stream: new restify.bunyan.RequestCaptureStream({
      level: bunyan.WARN,
      streams: streams
  })
}

streamsWithCapture.push(requestCaptureStream);
streamsWithCaptureFirst = [requestCaptureStream].concat(streams);

let logger = bunyan.createLogger({
    name: 'My Service',
    serializers: bunyan.stdSerializers,
    //testing this raw type, seems to have no effect
    type: 'raw',
    // without requestCaptureStream in array,
    // this writes info and error, but ignores debug
    //streams: streams });

    // with requestCaptureStream in array,
    // this writes the info and debug, but no error
    //streams: streamsWithCapture});

    // with just requestCaptureStream in array
    // this writes the info and debug, but no error
    streams: [ requestCaptureStream ]});

    // with requestCaptureStream first in array
    // this writes all of the messages, but writes
    // the info message twice
   // streams: streamsWithCaptureFirst });
    
logger.info('this is info');
logger.debug('this is debug');

if (process.argv[2] === 'error') {
  logger.error('error');
}
```

I'm testing my example app with:

    nodemon index.js error | bunyan

I'm sure that I've missed something obvious, but any input is appreciated before I start diving into the source for RequestCaptureStream.

Thanks!

**Comment by yunong** [comment]
I did some digging, and it looks like your example is not working because of two issues.

1) you're not adding a `req_id` field to your logging statements. The request by default expects you to always have a req_id field for each logging statement, unless you override it with `dumpDefault: true` as part of the arguments. See here: https://github.com/restify/node-restify/blob/5.x/lib/bunyan_helper.js#L69-L72

2) The reason you don't see the error log is that the request capture stream does not actually log out the error statement that triggers the dump of the logging records. Notice here: https://github.com/restify/node-restify/blob/5.x/lib/bunyan_helper.js#L182 that we only write the log statement to the ring buffer if it's level didn't trigger the dump. If it did, then we only dump out what is currently in the ring, and not that statement. If you add an additional stream to your logger, say `process.stdout`, you will see the error log.

2) might reasonable be a bug -- though I think the reason we don't dump it might be that we don't want to have 2 identical error records. Most people just redirect both the RCS and regular streams to the same file, and this would result in 2 error records.

**Comment by gerrard00** [comment]
Thank you for looking into this.

1) This is just a side effect of my simplifying the code to post here. In the real app the req_id is included in the log messages and I see that in the output.

2) I'm a bit confused, because that would lead me to believe that the scenario that I labelled "streamsWithCaptureFirst " would have worked. In that case, I created a logger with the  RequestStream as well as the real streams that I want to target. Unfortunately, that led to the info level messages which would be logged even without the dump being triggered being duplicated in the output.

I need my app to write the logs to multiple streams, including a stream which is directed to Logentries (using https://github.com/logentries/le_node). I'm not sure how to accomplish that based on the results of my testing. My naive expectation was that the RequestCaptureStream would act as a facade for all of my real streams.

I'll try to dig into the RequestCaputreStream code a bit and maybe things will become clearer.

**Comment by yunong** [comment]
there's definitely a bug in RCS where we're not capturing the logging record that *triggers* the dump of all records. That's what causing  your `error` record to not come through. The reason 'streamswithcapturefirst' works is because the records you're seeing are coming from both the RCS and stdout.

Effectively you're doing something like this

records -> RCS && stdout. But since there's a bug in RCS where the triggering record (in this case the error record) doesn't get written out, then you should expect to see info and debug from the RCS and only info and error for stdout (since the logger by default logs at info). The order of the streams has no effect on the output.

Does this make sense? Your code is working as expected, but due to a bug in RCS not logging out the triggering record (in this case error), and the fact that you're conflating the sink of your records, it might appear from inspection that RCS is not working.

**Comment by gerrard00** [comment]
That makes sense. If I can get RCS to dump the triggering record, then my original code using only the RCS with my logger should work correctly. I'll take a look at making that fix after the holiday.

**Comment by yunong** [comment]
fixed via 5425d4d. Thanks @gerrard00 for your contribution.
